source/MaterialXGenShader/ShaderStage.h

```diff
@@ -0,0 +1,303 @@
+
+#include <MaterialXCore/Library.h>
+#include <MaterialXCore/Node.h>
+#include <MaterialXGenShader/Syntax.h>
+#include <MaterialXGenShader/ShaderGraph.h>
+#include <MaterialXGenShader/GenOptions.h>
+
+#include <queue>
+#include <sstream>
+#include <unordered_map>
+
+/// Macro for being/end of statements to be picked up by a given shader stage.
+/// For shaders that are multi-stage all code generation statements adding code 
+/// to the shader should be wrapped inside such begin/end stating its target.
+#define BEGIN_SHADER_STAGE(stage, name) if (stage.getName() == name) {
+#define END_SHADER_STAGE(stage, name) }
+
+/// Private macro to name the main shader stage.
+/// This is only used to handle static initialization order issues.
+/// API users should use the string identifyer MaterialX::MAIN_STAGE
+/// instead of this macro.
+#define _MAIN_STAGE_NAME "main"
+
+namespace MaterialX
+{
+
+/// Identifier for main shader stage.
+extern const string MAIN_STAGE;
+
+class VariableBlock;
+using VariableBlockPtr = std::shared_ptr<VariableBlock>;
+using VariableBlockMap = std::unordered_map<string, VariableBlockPtr>;
+
+/// A block of variables in a shader stage
+class VariableBlock
+{
+  public:
+    VariableBlock(const string& name, const string& instance) 
+        : _name(name)
+        , _instance(instance)
+    {}
+
+    /// Get the name of this block.
+    const string& getName() const { return _name; }
+
+    /// Get the instance name of this block.
+    const string& getInstance() const { return _instance; }
+
+    /// Return true if the block has no variables.
+    bool empty() const { return _variableOrder.empty(); }
+
+    /// Return the number of variables in this block.
+    size_t size() const { return _variableOrder.size(); }
+
+    /// Return a variable by index.
+    ShaderPort* operator[](size_t index) { return _variableOrder[index]; }
+
+    /// Return a variable by index.
+    const ShaderPort* operator[](size_t index) const { return _variableOrder[index]; }
+
+    /// Return a variable by name. Throws exception if
+    /// no variable is found by the given name.
+    ShaderPort* operator[](const string& name);
+
+    /// Return a variable by name. Throws exception if
+    /// no variable is found by the given name.
+    const ShaderPort* operator[](const string& name) const;
+
+    /// Return a variable by name. Returns nullptr if
+    /// no variable is found by the given name.
+    ShaderPort* find(const string& name);
+
+    /// Return a variable by name. Returns nullptr if
+    /// no variable is found by the given name.
+    const ShaderPort* find(const string& name) const;
+
+    /// Add a new shader port to this block.
+    ShaderPort* add(const TypeDesc* type, const string& name, ValuePtr value = nullptr);
+
+    /// Add an existing shader port to this block.
+    void add(ShaderPortPtr port);
+
+  private:
+    string _name;
+    string _instance;
+    std::unordered_map<string, ShaderPortPtr> _variableMap;
```

Comment:
VariableBlocks (published uniforms/input/output) use the same ShaderPort class as used on ShaderNode/ShaderGraph internally. Removes having two representations for the same data, and ShaderPort instances can be shared from internal graph to public VariableBlocks.

Reply:
Nice to see this!